smartcontract/service/native/oracle.go

```diff
@@ -0,0 +1,392 @@
+/*
+ * Copyright (C) 2018 The ontology Authors
+ * This file is part of The ontology library.
+ *
+ * The ontology is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * The ontology is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with The ontology.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package native
+
+import (
+	"encoding/hex"
+	"encoding/json"
+	"math/big"
+
+	cstates "github.com/ontio/ontology/core/states"
+	scommon "github.com/ontio/ontology/core/store/common"
+	"github.com/ontio/ontology/errors"
+	"github.com/ontio/ontology/core/genesis"
+	"github.com/ontio/ontology/smartcontract/service/native/states"
+)
+
+const (
+	//function name
+	CREATE_ORACLE_REQUEST = "createOracleRequest"
+	SET_ORACLE_OUTCOME = "setOracleOutcome"
+	SET_ORACLE_CRON_OUTCOME = "setOracleCronOutcome"
+	CHANGE_CRON_VIEW = "changeCronView"
+
+	//keyPrefix
+	UNDO_TXHASH         = "UndoTxHash"
+	ORACLE_NUM          = "OracleNum"
+	REQUEST             = "Request"
+	OUTCOME_RECORD      = "OutcomeRecord"
+	FINAL_OUTCOME       = "FinalOutcome"
+	CRON_VIEW           = "CronView"
+	CRON_OUTCOME_RECORD = "CronOutcomeRecord"
+	FINAL_CRON_OUTCOME  = "FinalCronOutcome"
+)
+
+func init() {
+	Contracts[genesis.OracleContractAddress] = RegisterOracleContract
+}
+
+func RegisterOracleContract(native *NativeService) {
+	native.Register(CREATE_ORACLE_REQUEST, CreateOracleRequest)
+	native.Register(SET_ORACLE_OUTCOME, SetOracleOutcome)
+	native.Register(SET_ORACLE_CRON_OUTCOME, SetOracleCronOutcome)
+	native.Register(CHANGE_CRON_VIEW, ChangeCronView)
+}
+
+func CreateOracleRequest(native *NativeService) error {
+	params := new(states.CreateOracleRequestParam)
+	err := json.Unmarshal(native.Input, &params)
+	if err != nil {
+		return errors.NewDetailErr(err, errors.ErrNoCode, "[createOracleRequest] Contract params Unmarshal error!")
+	}
+
+	//check witness
+	err = validateOwner(native, params.Address)
+	if err != nil {
+		return errors.NewDetailErr(err, errors.ErrNoCode, "[createOracleRequest] validateOwner error!")
+	}
+
+	if params.OracleNum.Cmp(new(big.Int).SetInt64(0)) == 0 {
+		params.OracleNum.SetInt64(1)
+	}
+
+	contract := native.ContextRef.CurrentContext().ContractAddress
+	txHash := native.Tx.Hash()
+	txHashBytes := txHash.ToArray()
+	txHashHex := hex.EncodeToString(txHashBytes)
+	undoRequests := &states.UndoRequests{
+		Requests: make(map[string]interface{}),
```

Comment:
It  seems you use Map as Set? If so, use `map[string]struct{}` instead `interface{}` or `map[string]bool`  to make it more readable.